### PR TITLE
Add function to programmatically write to .env files

### PR DESCRIPTION
**Pull request by Berlioz**
Q1) I'm guessing it's not intentional for Functions params to only be allowed names that fit the current key format validation (i.e start with an uppercase letter or underscore). Should we coerce param names to start uppercase, hide them inside a reserved prefix, or do something else?

Q2) Is the file pointed to by FUNCTIONS_EMULATOR_DOTENV guaranteed to exist when running the emulator? If not, this code as written can cause writes to "real" dotenv files when running the emulator, which would be...bad.

### CODE REVIEW

File: src/functions/env.ts [hunk 1]
```diff
@@ -240,6 +252,62 @@ export function hasUserEnvs({
...
 }
 
+/**
+ * Write new environment variables into a dotenv file.
+ *
+ * Identifies one and only one dotenv file to touch using the same rules as loadUserEnvs().
+ * It is an error to provide a key-value pair which is already in the file.
+ * Not actually implemented yet.
+ */
+export function writeUserEnvs(toWrite: Record<string, string>, envOpts: UserEnvsOpts) {
```
Comment (inlined) [inline src/functions/env.ts:262]: I have a user experience question that i don't have a good answer for and would like to tag @taeold into here.

Should we actually be ever writing back to a project alias file or specifically a per-project file? E.g. if the alias is `dev` and we write back to that file, we're going to modify all dev projects for a codebase.

**Review by inlined — APPROVED** [review]
Fundamental devx Qs, but we can fix them with another PR if necessary. The interface of the method wouldn't change and you should feel fine building upon it.

File: src/functions/env.ts [hunk 2]
```diff
@@ -240,6 +252,62 @@ export function hasUserEnvs({
...
+  }
+}
+
+function createEnvFile(envOpts: UserEnvsOpts) {
+  let fileToWrite: string;
+  if (envOpts.isEmulator) {
+    fileToWrite = FUNCTIONS_EMULATOR_DOTENV;
+  }
+  fileToWrite = envOpts.isEmulator
+    ? FUNCTIONS_EMULATOR_DOTENV
```
Comment (taeold) [inline src/functions/env.ts:293]: logic here seems redundant. I don't think we need the first if block.

File: src/functions/env.ts [hunk 3]
```diff
@@ -240,6 +252,62 @@ export function hasUserEnvs({
...
+ * Identifies one and only one dotenv file to touch using the same rules as loadUserEnvs().
+ * It is an error to provide a key-value pair which is already in the file.
+ * Not actually implemented yet.
+ */
+export function writeUserEnvs(toWrite: Record<string, string>, envOpts: UserEnvsOpts) {
+  const { functionsSource, projectId, projectAlias, isEmulator } = envOpts;
+  let envFiles = findEnvfiles(functionsSource, projectId, projectAlias, isEmulator);
+  if (envFiles.length === 0) {
+    createEnvFile(envOpts);
+    envFiles = findEnvfiles(functionsSource, projectId, projectAlias, isEmulator);
```
Comment (taeold) [inline src/functions/env.ts:267]: nit* Maybe `createEnvFile` can return the name of the file it created. Then we can write:

```js
  envFiles = [createEnvFile(envOpts)]
```

File: src/test/functions/env.spec.ts [hunk 4]
```diff
@@ -282,6 +282,161 @@ FOO=foo
...
+    });
+
+    it("touches .env.projectId if there are no .env files and project alias is not available", () => {
+      env.writeUserEnvs({}, { projectId: "project", functionsSource: tmpdir });
+      expect(!!fs.statSync(path.join(tmpdir, ".env.project"))).to.be.true;
+    });
+
+    it("throws if asked to write a key that already exists in the most specific .env", () => {
+      createEnvFiles(tmpdir, {
+        [".env.alias"]: "FOO=foo",
```
Comment (taeold) [inline src/test/functions/env.spec.ts:319]: q: why do we have brackets around the string key?

**Comment by inlined** [comment]
RE .env.project and .env.alias, do we support both in the same deploy (I thought not...). Here it occurred to me that it might be nice to have it in the project file because we want to override my dev project, not all dev projects?

File: src/functions/env.ts [hunk 5]
```diff
@@ -240,6 +252,62 @@ export function hasUserEnvs({
...
+
+function createEnvFile(envOpts: UserEnvsOpts) {
+  let fileToWrite: string;
+  if (envOpts.isEmulator) {
+    fileToWrite = FUNCTIONS_EMULATOR_DOTENV;
+  }
+  fileToWrite = envOpts.isEmulator
+    ? FUNCTIONS_EMULATOR_DOTENV
+    : `.env.${envOpts.projectAlias || envOpts.projectId}`;
+  logger.debug(`No .env file detected to write to; creating ${fileToWrite}`);
```
[the remark this answers is not in the document]
Comment (Berlioz) [reply src/functions/env.ts]: Agreed.

File: src/functions/env.ts [hunk 6]
```diff
@@ -240,6 +252,62 @@ export function hasUserEnvs({
...
+  }
+
+  const currentEnvs = loadUserEnvs(envOpts);
+  for (const k of Object.keys(toWrite)) {
+    validateKey(k);
+    if (currentEnvs.hasOwnProperty(k)) {
+      throw new KeyValidationError(
+        k,
+        `Attempted to write param-defined key ${k} to .env files, but it was already defined.`
+      );
```
[the remark this answers is not in the document]
Comment (Berlioz) [reply src/functions/env.ts]: Done.

File: src/functions/env.ts [hunk 7]
```diff
@@ -240,6 +252,62 @@ export function hasUserEnvs({
   return findEnvfiles(functionsSource, projectId, projectAlias, isEmulator).length > 0;
 }
 
+/**
+ * Write new environment variables into a dotenv file.
+ *
+ * Identifies one and only one dotenv file to touch using the same rules as loadUserEnvs().
+ * It is an error to provide a key-value pair which is already in the file.
+ * Not actually implemented yet.
```
[the remark this answers is not in the document]
Comment (Berlioz) [reply src/functions/env.ts]: Oops. Fixed.

File: src/test/functions/env.spec.ts [hunk 8]
```diff
@@ -282,6 +282,161 @@ FOO=foo
...
+        fs.statSync(tmpdir);
+      }).to.throw();
+    });
+
+    it("touches .env.projectAlias if there are no .env files and project alias is available", () => {
+      env.writeUserEnvs(
+        {},
+        { projectId: "project", projectAlias: "alias", functionsSource: tmpdir }
+      );
+      expect(!!fs.statSync(path.join(tmpdir, ".env.alias"))).to.be.true;
```
[the remark this answers is not in the document]
Comment (Berlioz) [reply src/test/functions/env.spec.ts]: Yes. Changed.

**Comment by inlined** [comment]
@taeold: @mbleigh and I were discussing this in the context of some long-term work and we think it might actually _make sense_ to have alias and project ID both. For example, maybe my company's codebase checks in the `.env.dev` that sets up most of the required state to the right values but leaves one or two essential fields undefined which a .env.thomas-test-project-id would provide.

Does that change your position on allowing alias + project ID (the latter would take precedence since many projects can have the same alias)? I'd like us to come to consensus here since this affects resolution and where we should have writes go.

**Comment by taeold** [comment]
@inlined I like that model a lot. Let me start off an API proposal to make that change.

I think the change in behavior from throwing an error if both `.alias` and `.projectId` env files exists to having clearly documented precedence order is arguably not a breaking change so I'm going to propose it as such.

**Pull request merged**